x-pack/plugins/security_solution/server/endpoint/utils/action_list_helpers.ts

```diff
@@ -0,0 +1,188 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Logger } from '@kbn/core/server';
+// eslint-disable-next-line @kbn/eslint/no-restricted-paths
+import type { SearchRequest } from '@kbn/data-plugin/public';
+import type * as estypes from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+import type { TransportResult } from '@elastic/elasticsearch';
+import { AGENT_ACTIONS_RESULTS_INDEX } from '@kbn/fleet-plugin/common';
+import { EndpointError } from '../../../common/endpoint/errors';
+import type { SecuritySolutionRequestHandlerContext } from '../../types';
+import type {
+  EndpointAction,
+  LogsEndpointAction,
+  ActionListApiResponse,
+  EndpointActionResponse,
+  LogsEndpointActionResponse,
+} from '../../../common/endpoint/types';
+import {
+  ACTIONS_SEARCH_PAGE_SIZE,
+  ACTION_REQUEST_INDICES,
+  ACTION_RESPONSE_INDICES,
+} from '../services/actions/constants';
+import { getDateFilters, logsEndpointActionsRegex } from '../services/actions/utils';
+import { catchAndWrapError } from './wrap_errors';
+import { doesLogsEndpointActionsIndexExist } from './yes_no_data_stream';
+import { GetActionDetailsListParam } from '../services/actions/action_list';
+import { ENDPOINT_ACTIONS_INDEX } from '../../../common/endpoint/constants';
+
+const queryOptions = {
+  headers: {
+    'X-elastic-product-origin': 'fleet',
+  },
+  ignore: [404],
+};
+
+export const getTimeSortedActionDetails = (
+  data: ActionListApiResponse['data']
+): ActionListApiResponse['data'] => {
+  return data.sort((a, b) => (new Date(b.startedAt) > new Date(a.startedAt) ? 1 : -1));
+};
+
+export const getActionsForAgents = async ({
+  actionTypes,
+  context,
+  elasticAgentIds,
+  endDate,
+  from,
+  logger,
+  size,
+  startDate,
+  userIds,
+}: GetActionDetailsListParam): Promise<{
+  actionIds: string[];
+  actionRequests: TransportResult<
+    estypes.SearchResponse<EndpointAction | LogsEndpointAction>,
+    unknown
+  >;
+}> => {
+  const additionalFilters = [];
+  if (actionTypes?.length) {
+    additionalFilters.push({
+      terms: {
+        'data.command': actionTypes,
+      },
+    });
+  }
+  if (userIds?.length) {
+    additionalFilters.push({ terms: { user_id: userIds } });
+  }
+  if (elasticAgentIds?.length) {
+    additionalFilters.push({ terms: { agents: elasticAgentIds } });
+  }
+
+  const dateFilters = getDateFilters({ startDate, endDate });
+  const baseActionFilters = [
+    { term: { input_type: 'endpoint' } },
+    { term: { type: 'INPUT_ACTION' } },
+  ];
+  const actionsFilters = [...baseActionFilters, ...dateFilters];
+
+  const hasLogsEndpointActionsIndex = await doesLogsEndpointActionsIndexExist({
+    context,
+    logger,
+    indexName: ENDPOINT_ACTIONS_INDEX,
+  });
+  const actionsSearchQuery: SearchRequest = {
+    index: hasLogsEndpointActionsIndex ? ACTION_REQUEST_INDICES : AGENT_ACTIONS_RESULTS_INDEX,
```

Comment:
Thanks for clarifyiung @pzl . I tested this out and I'm wondering what should we do when there's no doc in `.fleet-actions` as in `.fleet-actions` index does not exist yet since no actions were created. There might also be the case that action was created but failed to deliver to fleet (written to endpoint index with an error endpoint response in the endpoint response index). Should we search on the existing endpoint index only in that case or completely ignore the search?

Reply:
@pzl we talked offline and agreed that we're going to search for actions only on the endpoint index and not on `.fleet-actions` index.